Switch rustdoc from `clean::Stability` to `rustc_attr::Stability`

jyn514 opened the pull request:
This gives greater type safety and is less work to maintain on the rustdoc end. It also makes rustdoc more consistent with rustc.
Noticed this while working on https://github.com/rust-lang/rust/issues/76998.

- Remove `clean::Stability` in favor of `rustc_attr::Stability`
- Remove `impl Clean for Stability`; it's no longer necessary

r? @GuillaumeGomez 
cc @petrochenkov

jyn514 committed:
Switch rustdoc from `clean::Stability` to `rustc_attr::Stability`

This gives greater type safety and is less work to maintain on the
rustdoc end.

jyn514 committed:
Remove unnecessary Clean impl

petrochenkov commented on compiler/rustc_attr/src/builtin.rs:
```diff
@@ -154,14 +155,27 @@ pub struct ConstStability {
+            (Self::Unstable { .. }, Self::Unstable { .. }) => Some(cmp::Ordering::Equal),
+            (Self::Stable { .. }, Self::Stable { .. }) => Some(cmp::Ordering::Equal),
+            (Self::Unstable { .. }, Self::Stable { .. }) => Some(cmp::Ordering::Less),
+            (Self::Stable { .. }, Self::Unstable { .. }) => Some(cmp::Ordering::Greater),
```
I'd keep this bit in rustdoc (and remove the `PartialOrd` impl from `StabilityLevel`?).
I don't know what rustdoc rendering wants when sorting by stability, but it probably isn't something generally applicable.

petrochenkov replied:
Also this can be simplified to `mem::discriminant(self).cmp(mem::discriminant(other))`.

jyn514 replied:
> Also this can be simplified to mem::discriminant(self).cmp(mem::discriminant(other)).

Does rustc guarantee that the order of the enums matches the order of the discriminants? I thought layout was unspecified.

petrochenkov replied:
Yes, discriminant values are specified, see https://doc.rust-lang.org/reference/items/enumerations.html?highlight=discriminant#custom-discriminant-values-for-fieldless-enumerations for enums without data, and https://github.com/rust-lang/rfcs/blob/master/text/2363-arbitrary-enum-discriminant.md#semantics for enums with data.

jyn514 replied:
TIL, thanks!

petrochenkov commented:
LGTM modulo the ordering bit.

jyn514 committed:
Move `PartialOrd` impl out of rustc

Rustdoc's ordering requirements are probably not relevant to the rest of
the compiler.

petrochenkov commented:
r? @petrochenkov @bors r+

bors commented:
:pushpin: Commit 96b0446b538a160c83ade3d9e5e9d2203868492d has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 96b0446b538a160c83ade3d9e5e9d2203868492d -->
<!-- homu: {"type":"Approved","sha":"96b0446b538a160c83ade3d9e5e9d2203868492d","approver":"petrochenkov"} -->
